### ISSUE TITLE
Add support for posting with Markdown syntax

### ISSUE THREAD

**Issue by cam-rod** [mastodon/mastodon#23981]
### Pitch

Requests to add support for posting with Markdown formatting has come up many times in the past, so this is a proposal to formalize the recent efforts. I'm not against other forms of markup formatting, but am focusing this issue specifically on Markdown.

My understanding is that Markdown support has not been fully approved or denied. The previous PR in #10787 (based on issue #853) seems to have been rejected due to issues in the implementation (and a somewhat derailed discussion). Issues #13544 and #21643 are similar, but cover subsections of this larger topic. _I consider this a continuation of the conversation in #18958._

Two main points stand out:

- **Is there general agreement to add Markdown support?** I think clearing up this question first will avoid the conflicts that appeared in the rejected PR.
- **If yes, how should Markdown be implemented?**
  - _Markdown flavour:_ Depending on the user, they might be coming from [CommonMark](https://commonmark.org/), the [GitHub-Flavoured derivative](https://github.github.com/gfm/), [Reddit](https://reddit.com/wiki/markdown) (where New Reddit is different from Old), [Misskey](https://misskey-hub.net/en/docs/features/mfm.html), [Redcarpet](https://github.com/vmg/redcarpet), or the [original spec](https://daringfireball.net/projects/markdown/syntax), to name a few. One of these needs to be selected to avoid issues from mixing them.
  - _Scope:_ Supporting all of Markdown would be a massive effort and massively increase risks of security vulnerabilities, so a subset of Markdown would be used. The rich text support added in #23913 is a good starting point.
  - _Locations:_ In addition to posts, should support be added to other places, like profiles? A small point, but it can matter (ex. supporting Markdown in profile metadata, but not display names)

### Motivation

With proper support for markup syntax, users can emphasize key points, format code snippets, create lists, or quote blog posts. It allows a message to be more expressive, similar to custom emojis. Adding support also means that users will no longer have to emulate the use of markup syntax:

```
I would like to make **this point clear.**
```

This means the content of the message is properly formatted into HTML, and can be interpreted by screen readers and other accessibility tools. It then displays as formatted on other servers, like Misskey and Akkoma.

**Comment by cam-rod** [comment]
Putting my personal opinion in a separate comment:

- I'm in favour of implementing Markdown for the reasons stated above.
- My implementation concept uses the parts of CommonMark that align with #23913. I would've preferred at least some size differential for header, keeping `<h1>` through `<h6>`  the same size, but all slightly bigger than regular text. but I think it's a good starting point otherwise. I'm divided on whether tables should be added, but currently lean no. 
- Supported locations would be posts and profile descriptions, but not display names or metadata (I think those should be kept relatively clean).

**Comment by crse** [comment]
I support for compability with Misskey-flavored Markdown, as it will minimize broken post like this:

![20230814_23 18 46 637_firefox_k30RxoXudS](https://github.com/mastodon/mastodon/assets/8639337/48e9dc36-b792-4b06-bcd8-fb250716f3da)
![20230814_23 18 59 216_firefox_VdgyaHyHJD](https://github.com/mastodon/mastodon/assets/8639337/1ecbc049-c03a-4d9b-83c7-b2a0f38261b8)

**Comment by quazar-omega** [comment]
Going for MFM seems overkill to me, I think Commonmark would fit quite well, also for ease of implementation by third-party apps

**Comment by awebeer256** [comment]
> Going for MFM seems overkill to me, I think Commonmark would fit quite well, also for ease of implementation by third-party apps

I would've said that this is too cautious, but only because I had not known anything at all about MFM until seeing the post before yours. That seems, uh, *extremely* excessive 😱.

**Comment by quazar-omega** [comment]
Yeah lol. To be fair, though, **plain** Commonmark is a bit too restrictive, I don't know which other specification is more featureful without being too much, but I think there are extensions that add support for the few missing things that could be added in (strikethrough, tables, superscript and subscript)

**Comment by Erlendhm** [comment]
IMO we'd get _so_ far with just
 
```
# heading
_italics_
**bold**
- Unordered lists
1. Ordered lists
> Block quote
[links](https://links.com)
code + code blocks
```

I like the idea of displaying h1-h6 as the same.

**Comment by CoffeeCoder1** [comment]
> IMO we'd get _so_ far with just
> 
> ```
> # heading
> _italics_
> **bold**
> - Unordered lists
> 1. Ordered lists
> > Block quote
> [links](https://links.com)
> code + code blocks
> ```
> 
> I like the idea of displaying h1-h6 as the same.

I think supporting ~strikethrough~ could also be quite useful.

```markdown
I think supporting ~strikethrough~ could also be quite useful.
```

**Comment by sloanlance** [comment]
I don't post on Mastodon as often as some people do, but when I do, it's fairly likely that I'll want to use Markdown formatting.  Fortunately, I found that the client named [Elk](/elk-zone/elk) works well _and_ it **_supports Markdown_**.

A recent post I wrote, using Markdown formatting for code segments, as seen in Elk: https://elk.zone/mastodon.social/@sloanlance/111537007812879508

**Comment by Reedyn** [comment]
> I don't post on Mastodon as often as some people do, but when I do, it's fairly likely that I'll want to use Markdown formatting. Fortunately, I found that the client named [Elk](/elk-zone/elk) works well _and_ it **_supports Markdown_**.

Solving it on the client side like Elk does is a whole different thing, that's just for you and it's not "real" and not without issues since it just parses unformatted posts and changes what it thinks could be markdown and displays it in HTML.

Not being able to actually author in Markdown which then gets published in formatted HTML to everyone.

**Comment by spartanatreyu** [comment]
A well defined set of markdown is nice but NEEDS one inclusion.

Most implementations (discord, reddit, stack overflow, etc...) seem to be adding an additional feature, the spoiler tag.

For example:

```md
My favourite part of Harry Potter is when ||spoiler text goes here.|| That part was awesome!
```

Which is rendered as:

```txt
My favourite part of Harry Potter is when ███████████████████████ That part was awesome!
```

The redacted content is replaced with the spoiler text once interacted with (typically when clicked on).

Spoiler tags aren't just for spoilers either, people are able to use them to block NSFW (not safe for work) content, and CW (content warnings), to prevent certain words/phrases/topics from unindentedly triggering people who have suffered some form of abuse.

JGM who co-created the markdown spec with Aaron Swartz said:

> spoilers is certainly in my top 3 of potential markdown extensions that people actually use and ask for.

**Comment by git-3000** [comment]
Markdown or BBcodes is desperately needed. 
It would also incentivize users to use image hosting websites to instead of uploading to through the instance. 

Markdown is so convenient

**Comment by Morethanevil** [comment]
Any news? Would love to have Markdown in Mastodon ☺️

**Comment by chrisale** [comment]
I just want the ability to link a word rather than paste in an entire URL. But either markdown or BBCode support would be extremely useful.

**Comment by MrAureliusR** [comment]
This really needs to be decided one way or the other. Myself and many others I know use Mastodon less than we otherwise would just because there is no Markdown support. Hell, give us ReStructuredText, give us LaTeX or Wikitext, anything! There cannot possibly be that much of a technical burden to changing this. I wish I had the spare time and I'd dig into it myself...

Honestly little things like these are holding Mastodon back significantly. We still can't even upload .opus or .ogg audio if I'm remembering correctly. Modern users want a modern interface with features that are standard in most other places they communicate.

**Comment by N-Upchurch** [comment]
> I just want the ability to link a word rather than paste in an entire URL. 

Yea, this and the ability to *communicate emphasis* seem like a pretty key features in a communications platform. I don't need headers or bullet points to communicate effectively, but the ability to emphasize certain words or phrases is very important, especially when the character count is restricted.

**Comment by ijackson** [comment]
### Summary of the situation.

As far as I can tell (it's not entirely clear):

 * On a technical level the proper way to do this in ActivityPub is to send a post whose source format is `text/markdown` with an appropriately rendered HTML.  Let's call that a "markdown post".
 * Since #23913, Mastodon does the right thing when other servers send markdown posts.
 * There is no support yet in Mastodon mainline for authoring posts in markdown.
 * Therefore this ticket is now about authoring posts in markdown.
 * Some (most?) Mastodon downstreams, and many other fediverse implementations, do support authoring posts in markdown
 * The upstream mastodon developers haven't said whether they want to support authoring posts in markdown.  That was the original question asked at the top of this ticket.
 * Authoring posts in mastodon is a highly-requested feature.

### Question for the maintainers

Do you support, in principle, the idea that users of Mastodon mainline should be able to author posts in markdown (or a subset thereof) ?

If the answer is "yes", please say so.  I'm sure there will be plenty of effort available for working out the details.

If the answer is "no", this ticket should be closed so that we can all stop wasting our time on it.  Also, in this case, a lot of people are going to be disappointed.  I think it would be a good idea in this case to give a clear and comprehensive rationale for the decision, so that the whole answer is in one place.  And, given recent governance changes in upstream Mastodon, it would be good to to be clear about whether that decision might be subject to review by the board of the new nonprofit entity, and if so how we would go about requesting such a review.

Thanks for your attention.

**Comment by mmahmoudian** [comment]
I was reading this thread and realized I'm not familiar with differences between Github Flavoured Markdown (GFM), Gitlab Flavoured Markdown (GLFM), and CommonMark (CM). I found this website that shows their differences in a table format:

https://www.ditig.com/markdown-parser-feature-comparison

For the sake of Mastodon, I believe GLFM is a better choice because it supports the following in addition to the [basic syntanx](https://www.ditig.com/markdown-parser-feature-comparison#basic-syntax) of Markdown that can be useful for posts in a micro-blogging platform like Mastodon:
1. Footnotes
2. Strikethrough
3. Math/LaTeX Support
4. Code Syntax Highlighting
5. Check Lists

As much as I like CommonMark, infortunately it does not support most of these features.

**Comment by quicoto** [comment]
Maybe @ClearlyClaire or @renchap would have more info about authoring posts with Markdown.

It's currently in https://joinmastodon.org/roadmap MAS-86 as "exploring".

**Comment by chrysn** [comment]
An aspect previously not brought up here so far is accessibility. Right now, software development posts on Mastodon frequently (25% on my timeline) contain screenshots from source code editors simply because it's the only way to get it to look like expected (i.e. with fixed width font and syntax highlighting); often, those are missing proper alt texts.

Any recent markup syntax (Markdown flavor or otherwise) provides means of declaring code, and would thus provide better accessibility. (And decreased media usage if that matters).

**Comment by ghosttie** [comment]
I think mastodon should have rich text posts, I don't care whether that's achieved with markdown or some other markup language - it's not like anyone's going to see the raw text anyway

**Comment by Eladriagon** [comment]
Has this been linked to the suggestion regarding previewing statuses before sending them?

While I don't think there's a direct dependency one way or the other (you can preview statuses without M🠋 to see how emojis and tags will render, and you can add M🠋 without a preview feature, as authoring M🠋 via plaintext is fairly common - take this box I'm typing in right now for example), I think it's still worth noting that the two independent features can be made to build upon one another, enriching the experience in both areas.

**Comment by thomaslule** [comment]
Im pretty sure that rich text would make my experience on mastodon worse. I love micro-blogging because it is simple, straightforward, short messages. It is far more easy for me to read my feed when there are only this kind of messages.

Sometimes, messages from other platforms reach my feed, they come with titles, quotes, hyperlinks, formatted text… they break the flow of reading and stand out in an awkward way.

**Comment by AkazaRenn** [comment]
That's the beauty of Mastodon, you can block anyone who posts contents you don't like and never see them again.

**Comment by ghosttie** [comment]
I don't think it's unreasonable to want to make a word bold or italics for emphasis

And people do it now with the scientific notation Unicode characters, which is really bad for accessibility

**Comment by quicoto** [comment]
I would love code blocks. People share code screenshots otherwise.

**Comment by MadokVaur** [comment]
> That's the beauty of Mastodon, you can block anyone who posts contents you don't like and never see them again.

Which misses the point of the recent additions to this comment thread entirely

The point is *not* to block --> content <-- that one doesn't want to read: that's elementary

The point is to hard-pass on those people who seem compelled to pretty-print their posts lavishly, as if this gives the --> content <-- greater weight

It does not

It merely takes up screen space with "Lookeee what I can doooo!!" bling to the detriment of the ideas expressed

**Comment by almereyda** [comment]
> Im pretty sure that rich text would make my experience on mastodon worse.

A less attention-seeking, user ergonomy respecting client could always choose to offer to opt to remove all formatting. Less is often more, but not in all cases for everyone. Why embracing diversity may also have something to offer.

I find Semphore with my Mastodon-compatible GoToSocial API a very calm environment. You might want to try Semaphore with your Mastodon account.

**Comment by N-Upchurch** [comment]
> The point is to hard-pass on those people who seem compelled to pretty-print their posts lavishly

My guy, if bold and italic text is too rich for your tastes, you might consider sticking to IRC. Also, any remotely competent copywriter understands the impact that _emphasis_ can have on a reader's interpretation of your prose.

Edit: I also note the utterly *lavish* formatting you have applied to your comment. A quote, italics, and even whatever these things are: `--><-- `.

**Comment by AkazaRenn** [comment]
> > That's the beauty of Mastodon, you can block anyone who posts contents you don't like and never see them again.
> 
> Which misses the point of the recent additions to this comment thread entirely
> 
> The point is *not* to block --> content <-- that one doesn't want to read: that's elementary
> 
> The point is to hard-pass on those people who seem compelled to pretty-print their posts lavishly, as if this gives the --> content <-- greater weight
> 
> It does not
> 
> It merely takes up screen space with "Lookeee what I can doooo!!" bling to the detriment of the ideas expressed

Doesn't change my point, block as you wish. Don't force your preference on others.

**Comment by segln** [comment]
You can also write custom CSS on your browser environment to force the tags generated from Markdown to appear as plain text.
This simple setup requires only a few characters of code.

**Comment by mmahmoudian** [comment]
It is interesting to see that some folks don't want/like formatting. If you don't want that, simply make your client to not apply formatting. Having formatting would drastically improve readability, but at the end of the day, your client is the one handling it. If it is implemented in a togglable way, those who doesn't want it can simply turn it off, and those who want it can thrn it on. But it is not logical to sacrifiece readability just because some portion of the community are devoted to plain-text.

Also, Mastodon is "microblogging" platform, but that's not entirely true. Many instances have changed the default character limit. Also because this is ActivityPub, one can get many long-form text in their feed (e.g., https://fe.disroot.org/notice/AziWdltrfqpqIGhWVM ).  Which brings us again to client-side settings. For example, the client I use on my phone only shows the first N characters of a post, and if the post is long, I can tap on the post and read the rest.

**Comment by Eladriagon** [comment]
I don't think that the group of people that don't want to view formatting in their feeds should be the determining factor on whether or not M🠋 / formatting is added to Mastodon.

It's trivial to convert M🠋 to HTML and then to strip HTML tags, so if you want to view posts without formatting, simply strip out the HTML from the status body.

The fact that, as of today, other fediverse clients can post statuses with HTML that appear and render correctly in Mastodon feeds *plus* people are bypassing the lack of formatting with Unicode and/or screenshots are two very strong reasons to adopt M🠋 (in some fashion) in Mastodon.

I personally feel that the debate is no longer "*Should we add this?*", but rather "*How do we add this?*".

**Comment by hasselmm** [comment]
> I would love code blocks. People share code screenshots otherwise.

This is important.

One of the key selling points of the Fediverse and especially Mastodon is its strong focus on being an inclusive place. We write alt texts, we use content warnings and all these things because we all want the Fediverse to be a better, a more inclusive place than any of the commercial networks.

Mastodon currently falls short on this goal, when it comes to code or any other text that requires proper formatting. People resort to screenshots because of that. This adds barriers on both sides, to audience and readers. It simply wastes time and resources, if you first have to screenshot, pretty-print and alt-text every single code snippet. It doesn't make sense to store code both as alt text and as wasteful image.

So if Mastodon shall be a place, where we educate people by sharing code, were we entertain people by sharing poems, then Mastodon strictly needs Markdown support. Not tomorrow, but last year.

**Comment by hasselmm** [comment]
> Im pretty sure that rich text would make my experience on mastodon worse. I love micro-blogging because it is simple, straightforward, short messages. It is far more easy for me to read my feed when there are only this kind of messages.
> 
> Sometimes, messages from other platforms reach my feed, they come with titles, quotes, hyperlinks, formatted text… they break the flow of reading and stand out in an awkward way.

It's good that you share your sentiments. Still there is an easy fix for your verify justified sentiments: Let Mastodon or a user script could filter any formatting. On the other hand: What is your fix for people who believe it is important to markup their toots? Or if we have to descide, and shall join your position: What makes your point of view more relevant than the others' point of view?

**Comment by thomaslule** [comment]
I came here only to share my concerns about this feature, like how a title can break the hierarchy of content of the feed, or how a quote/code block can stand out too much and disturb the reading, in the hope that maybe those concerns may be taken in account somehow if rich text is added (for example in the choice of supported markups or how they are displayed).

I dont think that suggesting that people should just add custom code to remove style is a great answer, because it is not that simple for everybody to accomplish that in every mastodon interfaces, and because it would result in: either raw markdown with code blocks displayed in plain text in the feed (I dont think anybody wants that), or in the worst of case a loss of information (like, if we simply strip the style, we dont know anymore that this part of a message was a quote). Also, blocking anyone who uses a new feature does not seem serious to me. Some of the answers just feel dismissive.

I dont know if my original message was badly phrased or if it is the usual culture of this place, but the interactions here are not really pleasant, I prefer to not participate again on the subject, all the best to you all! ❤️

**Comment by BoD** [comment]
It seems pretty evident that if the feature is added, a setting to remove formatting from other posts must also be added at the same time.

**Comment by chrysn** [comment]
May I suggest tracking a user-configurable view setting for removal of formatting in a separate issue?

It is apparent that there is desire to do this, but the need for this is already there: Unless I'm living in an extreme non-bubble, a good portion of ActivityPub messages in a Mastodon user's timeline already stem from non-Mastodon instances, and plenty of them already produce messages that contain HTML formatting.

**Comment by BoD** [comment]
@chrysn I agree with the benefits of tackling these 2 separately, but I'm afraid doing can be a friction to adopt this feature. Because as we can see from the above discussions, some people find formatting undesirable. If both abilities are bundled together there is no objection.

**Comment by jargon4220** [comment]
> I came here only to share my concerns about this feature, like how a title can break the hierarchy of content of the feed, or how a quote/code block can stand out too much and disturb the reading, in the hope that maybe those concerns may be taken in account somehow if rich text is added (for example in the choice of supported markups or how they are displayed).
> 
> I dont think that suggesting that people should just add custom code to remove style is a great answer, because it is not that simple for everybody to accomplish that in every mastodon interfaces, and because it would result in: either raw markdown with code blocks displayed in plain text in the feed (I dont think anybody wants that), or in the worst of case a loss of information (like, if we simply strip the style, we dont know anymore that this part of a message was a quote). Also, blocking anyone who uses a new feature does not seem serious to me. Some of the answers just feel dismissive.
> 
> I dont know if my original message was badly phrased or if it is the usual culture of this place, but the interactions here are not really pleasant, I prefer to not participate again on the subject, all the best to you all! ❤️

One can always choose not to use an existing feature but one cannot choose to use a feature that does not exist. It feels necessary to start with this elementary concept since you seem to either deny or misunderstand it.

If your feed is being polluted by what you consider an abuse of formatting the solution is to filter those styles or mute the users in question. Just as you dislike formatting a significant number of users desire it and the ActivityPub ecosystem is already teeming with formatted text. Furthermore the current workaround of uploading screenshots just to display structured text disrupts the feed just as much if not more. Your grievances provide no valid grounds for obstructing the implementation of Markdown.

It is worth noting that clients retain full control to process and render or strip formatted text as they see fit. I fail to see the provenance of your argument here. It suggests a fundamental lack of understanding regarding the technical architecture at play. Your concerns are quite frankly moot in this context.

Must we really reiterate the basic principle that if one desires a specific outcome one must find the means to achieve it? If you are so adamant about stripping formatting then the onus is on you to implement a solution. If you find that task too difficult the logical steps are to seek assistance or simply accept the status quo rather than expecting the entire platform to stagnate for your convenience.

Every single drawback you have cataloged is vastly superior to the atrocious current practice of uploading text as screenshot images to bypass limitations. Moreover a loss of information resulting from user side stripping where the raw source remains available is infinitely preferable to having no method to convey that information in the first place.

In summary your points lack productivity. Frankly it appears you either fail to grasp the technical nuances being discussed here or you are simply refusing to concede out of misplaced pride.

**Comment by chrysn** [comment]
> If both abilities are bundled together there is no objection.

This could also be achieved by the maintainers promising that this won't be merged that to-be-created issue is resolved. (On the PR side, this can be modeled as one PR blocking the other; IIUC, GitHub issues don't support this directly).

Split discussions could help get the right people in the right discussions: Personally, I'd love to contribute to discussing how rich formatting can be implemented in the GUI, but am about 2 rants about how the users need to be able to opt out (or longwinded rebuttals thereof) of it away from just unsubscribing from the thread and declaring it a lost cause.

**Comment by BoD** [comment]
@chrysn Fair enough 👍 
My point wasn't too much about the details of how these 2 features are designed / discussed / and worked on, but more about the assurance that formatting options are released at the same time as the ability to strip them.

Is there already a ticket about stripping formatting? If not now may be a good time to create it.

**Comment by BoD** [comment]
@gestate-8 I think we can put this debate to rest (especially since the person indicated they no longer wish to discuss it), if we have the ability to ignore/strip formatting. This addresses the concern, and we can move on.

**Comment by LokiMidgard** [comment]
I created #36990 in hopes that some of the discussion — specifically around how to handle formatting when it is not desired — can be held productively there. This includes addressing how Mastodon handles formatted posts coming from non-Mastodon ActivityPub instances.

That way, this issue can focus more on ***how*** Mastodon should contribute rich text content to the Fediverse.

Discussing both features separately will hopefully be more constructive. I don't think people who want to add formatting will help much with a feature for removing formatting, and people who want to get rid of formatting are likely not interested in what syntax/features should be supported.
